Enumerate Docker containers as new Evidences to be processed

rgayon opened the pull request:
* Add a new DockerContainer evidence type as well as processors to mount them
* Add a DockerContainersEnumerationTask to generate DockerContainer evidences to be processed
* Add DockerContainer as an input Evidence type to some tasks (Only those that are 'quick')
* Fix a couple string formatting issues (Py3 hates passing None to {0:s})

Reminder: For every found Docker container, all Tasks that can have it as input will be run. For every run, all preprocessing for every container will be re-run (attaching/mounting disks/mounting the container FS). This mean if we have N tasks taking DockerContainer Evidence as input, and M docker containers, a google cloud disk will be attached/detached M*N times

rgayon commented on turbinia/evidence.py:
```diff
@@ -615,3 +616,43 @@ def __init__(self, module_list=None, profile=None, *args, **kwargs):
+  """
+
+  # ABSOLUTELY NO LEADING / HERE
+  DEFAULT_DOCKER_DIRECTORY_PATH = 'var/lib/docker'
```
you mean being passed as a flag of turbiniactl?
I could devise a small but the directory is pretty much always that, unless configuration says otherwise. 
https://github.com/moby/moby/blob/51f27b1defad7418675567cba6aa6c95efebf739/cmd/dockerd/config_common_unix.go#L28

I guess I could also add something simple that could check if /etc/docker/daemon.json says something different

aarontp replied:
Ok, yeah, checking the json file for the path sounds good.  We can do that in a separate PR though since it's usually the same.  Maybe file an issue for that?

rgayon replied:
opened https://github.com/google/turbinia/issues/479

aarontp commented on turbinia/workers/docker.py:
```diff
@@ -0,0 +1,108 @@
+log = logging.getLogger('turbinia')
+
+
+class DockerContainersEnumerationTask(TurbiniaTask):
```
Can we also add a test here?  It should be relatively easy to test Tasks now, there is a `TestTurbiniaTaskBase` that you can inherit from and pass in the right evidence, etc.  Here's an example:
https://github.com/google/turbinia/blob/master/turbinia/workers/bulk_extractor_test.py

rgayon replied:
Done

aarontp replied:
Thanks!

rgayon commented on turbinia/evidence.py:
```diff
@@ -615,3 +616,43 @@ def __init__(self, module_list=None, profile=None, *args, **kwargs):
+
+    self.context_dependent = True
+
+  def _preprocess(self, _):
```
yup. Removed

rgayon commented on turbinia/processors/docker.py:
```diff
@@ -0,0 +1,90 @@
+  log.info(
+      'Using docker_explorer to mount container {0:s} on {1:s}'.format(
+          container_id, container_mount_path))
+  # TODO(aarontp): Remove hard-coded sudo in commands:
```
done

rgayon commented on turbinia/processors/docker.py:
```diff
@@ -0,0 +1,90 @@
+          container_id, container_mount_path))
+  # TODO(aarontp): Remove hard-coded sudo in commands:
+  # https://github.com/google/turbinia/issues/73
+  de_paths = [
```
I can do my own which by looking into PATH :)

rgayon commented on turbinia/processors/docker.py:
```diff
@@ -0,0 +1,90 @@
+
+  try:
+    subprocess.check_call(mount_cmd)
+  except Exception as e:
```
done

rgayon commented on turbinia/processors/docker.py:
```diff
@@ -0,0 +1,90 @@
+  try:
+    subprocess.check_call(mount_cmd)
+  except Exception as e:
+    raise TurbiniaException('Could not mount container: {0!s}'.format(e))
```
done

rgayon commented on turbinia/workers/docker.py:
```diff
@@ -0,0 +1,108 @@
+      for container_info in containers_info:
+        container_id = container_info.get('container_id')
+        found_containers.append(container_id)
+        container_evidence = DockerContainer(container_id=container_id)
```
https://github.com/google/turbinia/blob/master/turbinia/workers/__init__.py#L251

It's magic ! =)

rgayon commented on turbinia/workers/docker.py:
```diff
@@ -0,0 +1,108 @@
+        'sudo', de_binary, '-r', docker_dir, 'list', 'all_containers'
+    ]
+    log.info('Running {0:s}'.format(' '.join(docker_explorer_command)))
+    json_string = subprocess.check_output(docker_explorer_command).decode(
```
by " trap the appropriate exceptions", you mean things like "invalid json" or the de.py command not working as expected?

aarontp replied:
Yeah, I think probably both of those.

rgayon replied:
DOne

aarontp commented on turbinia/workers/__init__.py:
```diff
@@ -168,14 +168,19 @@ def close(self, task, success, status=None):
+            task.output_manager.save_evidence(evidence, self)
+        else:
+          self.log(
+              'Evidence {0:s} has missing file at source_path {1:s} so '
```
Sorry, I didn't catch this before, but can you change this to `{1!s}` so that in case the `evidence.source_path` is None that python3 won't complain about type mismatches?

rgayon replied:
done

rgayon commented:
So I tests on a googleclouddiskembedded evidence, where the disk is a GKE similar installation with 27 containers, 1 of them having compromised Wordpress & bad SShdConfigFile:

```
$ turbiniactl status -r 2c0c3447924645f98f58d566ce114b05 > /tmp/report
$ grep -v "Extracted 0" /tmp/report  | grep -v "No Wordpress"


# Turbinia report 2c0c3447924645f98f58d566ce114b05
* Processed 140 Tasks for user romaing

# High Priority Tasks
* WordpressAccessLogAnalysisTask: Wordpress access logs found (install, theme_edit)

# Successful Tasks
* FileArtifactExtractionTask: Extracted 1 new SshdConfigFile artifacts
* FileArtifactExtractionTask: Extracted 27 new GKEDockerContainerLogs artifacts
* DockerContainersEnumerationTask: Found 27 containers: bf830d68d71fb129378da9da4759a803a2a1823ed08ff72ccd8e14f766840132 6bc2a4c8051e68221f9dd408bbb0b81fa97384c9dedb1b32d53bf567df81db34 4a26bbe260b3681483436ee64e11d167712fad620ea6e5174b2662691b15e834 712d2a939ce22e7279025fb97d06442b7169046876a1a95f31786891be2d1e5d 01d2d0c1e3f6ea4818ce07882ae51e776517d78762e9b035919396dc84bb8253 7f644964d4097506be17c2481c609b2e566851339f99582d50d01a4d614f9673 ebc88302deba4854568eea9567b0bd1ee01ca8cbb8f5891aba1b8e0de242ff78 796197bb8a36f3e594a9ee452ed76455d23f1052f9f85f016e75be2887932ab7 e917f8eeab0ce0292115ad3cfc774016370f1e2164b78f0768eeb0a47a7ef71e 6393c90536c41178dd1489d0f8192a03a435f2c758847148f7b04642aba73116 2b1de554fe3f1068fd29ae156b033c2c866970b4a65afb2d9425278906bfc026 fc88af48094fc431e5fbf98c28fd0a6a5b57211c2b18f139835bbd6b1ceaac20 4335c9303a78c6172a9468908ff0ab51e9b5a8366569c42e3115aa69a35f7a44 1e9796e72d96e6733873adfc9535df64a40acc6f178aa13c8298f44a0c63d84a d4efa468e94cab28009f5a9804f5cf5ea580e8a069f62e1f8342e6d3a13ec023 c65720d0e452f610ee8f1eefd4f2e057bd6d8eec17ab3779cd6287c968378231 37c48eea7f95717609395f89af0ce87c32d9089a8a6f1306c1a73d253d0bb57d 925e2b4a2b2f6964619fb710d480293bd408c1203612b153ffb591d06a5f5615 6cf318d29ffec371df7fb212da08347890b37a3b65c4a16adc44200824537c68 ea92e82d2edda85ef15edde80793063ee01f8284eb19712aa9ad0c02d6f75b74 c8e3714b5efa123c40efe3f64e59bc3a70897f4ceffdd63eb86e5f873946ec57 569e7ac6f1188a18114d5836cfcf0070f276188371463d6e8b6ce89c82dcac69 9ed35421875db37115a7175de3bb1642cff35f59acd9[...]

# Failed Tasks
* None

# Scheduled or Running Tasks
* None

```
